### PR TITLE
Web Inspector: Color picker should allow picking a color from any pixel on screen

### PR DESCRIPTION
**Pull request by patrickangle**
#### 279b6e75d10ebaa7c646aab86685736c8b93e744
<pre>
Web Inspector: Color picker should allow picking a color from any pixel on screen
<a href="https://bugs.webkit.org/show_bug.cgi?id=124357">https://bugs.webkit.org/show_bug.cgi?id=124357</a>
rdar://15469621

Reviewed by Devin Rousso.

Add the long-missing &quot;eyedropper&quot; for picking a color from the screen to the Color Picker in Web Inspector. This
implementation supports both sRGB color space as well as Display P3 (the two colorspaces currently supported by the Web
Inspector frontend). The existing value format and gamut is preserved as best as possible while not clamping the color
(unless interacting with a color swatch/picker that does not allow changing the format). For example, on macOS the color
picker will return a Display-P3 color on supported displays, but the color itself may be representable in sRGB. If the
existing color you are overwriting is already in sRGB, that is preserved by converting the system&apos;s Display-P3 color
into sRGB. If the sampled color can not be represented in sRGB, we update the CSS value to support the new wider gamut
of the selected color in order to provide the best fidelity in color matching.

The picker is implemented in two places. The first is the Color Picker popovers used in Web Inspector. The second place
is as an Option-Click action for inline swatches to allow you to quickly begin color selection without opening the
picker itself.

* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:
* Source/WebInspectorUI/UserInterface/Images/Pipette.svg: Added.

* Source/WebInspectorUI/UserInterface/Models/Color.js:
(WI.Color.prototype.fromStringBestMatchingSuggestedFormatAndGamut):
- New convencience &quot;constructor&quot; for creating a color from a String while attempting (or forcing) preservation of an
existing format and gamut.

* Source/WebInspectorUI/UserInterface/Views/ColorPicker.css:
(.color-picker  &gt; .color-inputs-wrapper):
(.color-picker  &gt; .color-inputs-wrapper  &gt; .color-inputs):
(.color-picker  &gt; .color-inputs-wrapper  &gt; .color-inputs  &gt; div):
(.color-picker  &gt; .color-inputs-wrapper  &gt; .color-inputs  &gt; div + div):
(.color-picker  &gt; .color-inputs-wrapper  &gt; .color-inputs input):
(.color-picker  &gt; .color-inputs-wrapper  &gt; .pick-color-from-screen):
(.color-picker  &gt; .color-inputs-wrapper  &gt; .pick-color-from-screen.active):
(.color-picker  &gt; .color-inputs): Deleted.
(.color-picker  &gt; .color-inputs  &gt; div): Deleted.
(.color-picker  &gt; .color-inputs  &gt; div + div): Deleted.
(.color-picker  &gt; .color-inputs input): Deleted.
* Source/WebInspectorUI/UserInterface/Views/ColorPicker.js:
(WI.ColorPicker.async pickColorFromScreen):
- Add new static method to picking a color from the screen (so that InlineSwatch can use it), and add a new Pipette icon
to begin the modal color picking mode.

* Source/WebInspectorUI/UserInterface/Views/InlineSwatch.js:
(WI.InlineSwatch.prototype._updateSwatch):
- Add Option-Click for color swatches to immediately enter the modal color picking mode.

* Source/WebCore/inspector/InspectorFrontendClient.h:
* Source/WebCore/inspector/InspectorFrontendClientLocal.h:
* Source/WebCore/inspector/InspectorFrontendHost.cpp:
(WebCore::InspectorFrontendHost::canPickColorFromScreen):
(WebCore::InspectorFrontendHost::pickColorFromScreen):
* Source/WebCore/inspector/InspectorFrontendHost.h:
* Source/WebCore/inspector/InspectorFrontendHost.idl:
* Source/WebKit/UIProcess/Inspector/RemoteWebInspectorUIProxy.cpp:
(WebKit::RemoteWebInspectorUIProxy::pickColorFromScreen):
(WebKit::RemoteWebInspectorUIProxy::platformPickColorFromScreen):
* Source/WebKit/UIProcess/Inspector/RemoteWebInspectorUIProxy.h:
* Source/WebKit/UIProcess/Inspector/RemoteWebInspectorUIProxy.messages.in:
* Source/WebKit/UIProcess/Inspector/WebInspectorUIProxy.cpp:
(WebKit::WebInspectorUIProxy::pickColorFromScreen):
(WebKit::WebInspectorUIProxy::platformPickColorFromScreen):
* Source/WebKit/UIProcess/Inspector/WebInspectorUIProxy.h:
* Source/WebKit/UIProcess/Inspector/WebInspectorUIProxy.messages.in:
* Source/WebKit/WebProcess/Inspector/RemoteWebInspectorUI.cpp:
(WebKit::RemoteWebInspectorUI::pickColorFromScreen):
* Source/WebKit/WebProcess/Inspector/RemoteWebInspectorUI.h:
* Source/WebKit/WebProcess/Inspector/WebInspectorUI.cpp:
(WebKit::WebInspectorUI::pickColorFromScreen):
(WebKit::WebInspectorUI::canPickColorFromScreen):
* Source/WebKit/WebProcess/Inspector/WebInspectorUI.h:
- Add plubming for picking a color from the screen.

* Source/WebKit/UIProcess/Inspector/mac/RemoteWebInspectorUIProxyMac.mm:
(WebKit::RemoteWebInspectorUIProxy::platformPickColorFromScreen):
* Source/WebKit/UIProcess/Inspector/mac/WebInspectorUIProxyMac.mm:
(WebKit::WebInspectorUIProxy::platformPickColorFromScreen):
- Use NSColorSampler to get the system &quot;pick a color from the screen&quot; UI.

* Source/WebKit/UIProcess/Inspector/gtk/RemoteWebInspectorUIProxyGtk.cpp:
(WebKit::RemoteWebInspectorUIProxy::platformPickColorFromScreen):
* Source/WebKit/UIProcess/Inspector/gtk/WebInspectorUIProxyGtk.cpp:
(WebKit::WebInspectorUIProxy::platformPickColorFromScreen):
* Source/WebKit/UIProcess/Inspector/win/RemoteWebInspectorUIProxyWin.cpp:
(WebKit::RemoteWebInspectorUIProxy::platformPickColorFromScreen):
* Source/WebKit/UIProcess/Inspector/win/WebInspectorUIProxyWin.cpp:
(WebKit::WebInspectorUIProxy::platformPickColorFromScreen):
* Source/WebKit/WebProcess/Inspector/gtk/WebInspectorUIGtk.cpp:
(WebKit::WebInspectorUI::canPickColorFromScreen):
* Source/WebKit/WebProcess/Inspector/mac/WebInspectorUIMac.mm:
(WebKit::WebInspectorUI::canPickColorFromScreen):
* Source/WebKit/WebProcess/Inspector/win/WebInspectorUIWin.cpp:
(WebKit::WebInspectorUI::canPickColorFromScreen):
- Add stubs for Windows/GTK implementations.

Canonical link: <a href="https://commits.webkit.org/251236@main">https://commits.webkit.org/251236@main</a>
git-svn-id: <a href="https://svn.webkit.org/repository/webkit/trunk@295147">https://svn.webkit.org/repository/webkit/trunk@295147</a> 268f45cc-cd09-0410-ab3c-d52691b4dbfc
</pre>

### CODE REVIEW

File: Source/WebCore/inspector/InspectorFrontendHost.cpp [hunk 1]
```diff
@@ -494,6 +497,45 @@ void InspectorFrontendHost::load(const String& path, Ref<DeferredPromise>&& prom
...
+
+    m_client->pickColorFromScreen([promise = WTFMove(promise)](const std::optional<WebCore::Color>& color) {
+        if (!color) {
+            promise->resolve();
+            return;
+        }
+
+        String serializedColor;
+        // FIXME: <webkit.org/b/241198> Inspector frontend should support all color function gamuts.
+        if (color->colorSpace() != ColorSpace::SRGB || color->colorSpace() != ColorSpace::DisplayP3) {
```
Comment (dcrousso) [inline Source/WebCore/inspector/InspectorFrontendHost.cpp:522]: AFAICT it seems like we have `serializationForCSS` variants for each `ColorSpace`.  Am I missing some deeper detail preventing us from just doing that?

File: Source/WebKit/UIProcess/Inspector/mac/RemoteWebInspectorUIProxyMac.mm [hunk 2]
```diff
@@ -241,6 +243,23 @@ - (BOOL)inspectorViewControllerInspectorIsUnderTest:(WKInspectorViewController *
...
+void RemoteWebInspectorUIProxy::platformPickColorFromScreen(CompletionHandler<void(const std::optional<WebCore::Color>&)>&& completionHandler)
+{
+    auto completionBlock = makeBlockPtr([completionHandler = WTFMove(completionHandler)](const std::optional<WebCore::Color>& color) mutable {
+        completionHandler(color);
+    });
+
+    NSColorSampler *sampler = [[NSColorSampler alloc] init];
+    [sampler showSamplerWithSelectionHandler:[completionBlock](NSColor *selectedColor) {
+        if (!selectedColor) {
+            completionBlock({ });
```
Comment (dcrousso) [inline Source/WebKit/UIProcess/Inspector/mac/RemoteWebInspectorUIProxyMac.mm]: NIT: `std::nullopt`?

File: Source/WebKit/UIProcess/Inspector/mac/RemoteWebInspectorUIProxyMac.mm [hunk 3]
```diff
@@ -241,6 +243,23 @@ - (BOOL)inspectorViewControllerInspectorIsUnderTest:(WKInspectorViewController *
         completionHandler(nullString());
 }
 
+void RemoteWebInspectorUIProxy::platformPickColorFromScreen(CompletionHandler<void(const std::optional<WebCore::Color>&)>&& completionHandler)
+{
+    auto completionBlock = makeBlockPtr([completionHandler = WTFMove(completionHandler)](const std::optional<WebCore::Color>& color) mutable {
```
Comment (dcrousso) [inline Source/WebKit/UIProcess/Inspector/mac/RemoteWebInspectorUIProxyMac.mm]: Why is this necessary?

File: Source/WebKit/UIProcess/Inspector/mac/WebInspectorUIProxyMac.mm [hunk 4]
```diff
@@ -590,6 +592,23 @@ - (void)inspectorViewController:(WKInspectorViewController *)inspectorViewContro
         completionHandler(nullString());
 }
 
+void WebInspectorUIProxy::platformPickColorFromScreen(CompletionHandler<void(const std::optional<WebCore::Color>&)>&& completionHandler)
+{
+    auto completionBlock = makeBlockPtr([completionHandler = WTFMove(completionHandler)](const std::optional<WebCore::Color>& color) mutable {
```
Comment (dcrousso) [inline Source/WebKit/UIProcess/Inspector/mac/WebInspectorUIProxyMac.mm]: ditto (Source/WebKit/UIProcess/Inspector/mac/RemoteWebInspectorUIProxyMac.mm:248)

File: Source/WebKit/UIProcess/Inspector/mac/WebInspectorUIProxyMac.mm [hunk 5]
```diff
@@ -590,6 +592,23 @@ - (void)inspectorViewController:(WKInspectorViewController *)inspectorViewContro
...
+        completionHandler(color);
+    });
+
+    NSColorSampler *sampler = [[NSColorSampler alloc] init];
+    [sampler showSamplerWithSelectionHandler:[completionBlock](NSColor *selectedColor) {
+        if (!selectedColor) {
+            completionBlock({ });
```
Comment (dcrousso) [inline Source/WebKit/UIProcess/Inspector/mac/WebInspectorUIProxyMac.mm]: ditto (Source/WebKit/UIProcess/Inspector/mac/RemoteWebInspectorUIProxyMac.mm:255)

File: Source/WebInspectorUI/UserInterface/Models/Color.js [hunk 6]
```diff
@@ -209,6 +209,74 @@ WI.Color = class Color
...
+            newColor.gamut = WI.Color.Gamut.SRGB;
+
+        // Non-sRGB gamuts can only be expressed in the Color Function format.
+        if (newColor.gamut !== WI.Color.Gamut.SRGB)
+            return newColor;
+
+        // Match as closely as possible the suggested format, and progressively adjust the format (e.g. ShortHEX -> HEX
+        // -> HEXAlpha) if an exact match would be lossy.
+        switch (suggestedFormat) {
+            case WI.Color.Format.Original:
```
Comment (dcrousso) [inline Source/WebInspectorUI/UserInterface/Models/Color.js]: Style: the `case` should be on the same column/indentation as the `switch`

File: Source/WebInspectorUI/UserInterface/Models/Color.js [hunk 7]
```diff
@@ -209,6 +209,74 @@ WI.Color = class Color
...
+
+            case WI.Color.Format.HEXAlpha:
+            case WI.Color.Format.RGBA:
+            case WI.Color.Format.HSLA:
+            case WI.Color.Format.ColorFunction:
+                newColor.format = suggestedFormat;
+                break;
+
+            default:
+                console.assert(false, "Should not be reached.", suggestedFormat);
```
Comment (dcrousso) [inline Source/WebInspectorUI/UserInterface/Models/Color.js]: missing a `break`

File: Source/WebInspectorUI/UserInterface/Views/ColorPicker.css [hunk 8]
```diff
@@ -76,23 +76,41 @@
...
     display: flex;
     align-items: center;
     width: 100%;
 }
 
-.color-picker > .color-inputs > div + div {
-    margin-inline-start: 4px;
+.color-picker > .color-inputs-wrapper > .color-inputs > div + div {
+    
 }
```
Comment (dcrousso) [inline Source/WebInspectorUI/UserInterface/Views/ColorPicker.css]: delete this?

File: Source/WebInspectorUI/UserInterface/Views/ColorPicker.js [hunk 9]
```diff
@@ -56,7 +58,34 @@ WI.ColorPicker = class ColorPicker extends WI.Object
         wrapper.appendChild(this._hueSlider.element);
         wrapper.appendChild(this._opacitySlider.element);
 
-        this._element.appendChild(this._colorInputsContainerElement);
+        let colorInputsWrapperElement = this._element.appendChild(document.createElement("div"));
+        colorInputsWrapperElement.classList.add("color-inputs-wrapper");
+        colorInputsWrapperElement.appendChild(this._colorInputsContainerElement);
+
+        if (InspectorFrontendHost.canPickColorFromScreen()) {
+            let pickColorElement = WI.ImageUtilities.useSVGSymbol("Images/Pipette.svg", "pick-color-from-screen", WI.UIString("Pick color from screen", "Color picker view tooltip for picking a color from the screen."));
```
Comment (dcrousso) [inline Source/WebInspectorUI/UserInterface/Views/ColorPicker.js:66]: rather than having `<svg>`, could we make this a `<div>` (or `<button>`) and have this?
```
width: 16px;
height: 16px;
background-image: url("../Images/Pipette.svg");
```

File: Source/WebInspectorUI/UserInterface/Views/ColorPicker.js [hunk 10]
```diff
@@ -56,7 +58,34 @@ WI.ColorPicker = class ColorPicker extends WI.Object
...
+            pickColorElement.style.width = "16px";
+            pickColorElement.style.height = "16px";
+
+            pickColorElement.role = "button";
+            pickColorElement.addEventListener("click", async (event) => {
+                pickColorElement.classList.toggle("active", true);
```
Comment (dcrousso) [inline Source/WebInspectorUI/UserInterface/Views/ColorPicker.js]:
```
pickColorElement.classList.add("active");
```

File: Source/WebInspectorUI/UserInterface/Views/ColorPicker.js [hunk 11]
```diff
@@ -56,7 +58,34 @@ WI.ColorPicker = class ColorPicker extends WI.Object
...
+                let pickedColor = await WI.ColorPicker.pickColorFromScreen({
+                    suggestedFormat: this.color.format,
+                    suggestedGamut: this.color.gamut,
+                    forceSuggestedFormatAndGamut: this._preventChangingColorFormats,
+                });
+                pickColorElement.classList.toggle("active", false);
```
Comment (dcrousso) [inline Source/WebInspectorUI/UserInterface/Views/ColorPicker.js]:
```
pickColorElement.classList.remove("active");
```

File: Source/WebInspectorUI/UserInterface/Views/ColorPicker.js [hunk 12]
```diff
@@ -68,6 +97,30 @@ WI.ColorPicker = class ColorPicker extends WI.Object
...
+    {
+        console.assert(InspectorFrontendHost.canPickColorFromScreen());
+
+        // There is a brief moment where the frontend page remains interactable before the backend actually begins the
+        // modal color picking mode. In order to avoid accidentally hovering an element and showing its highlight on the
+        // page and not being able to hide the highlight while selecting a color, make the document inert so that even
+        // immediate mouse movement doesn't accidentaly cause any highlighting to occur.
+        document.body.inert = true;
+
+        let pickedColorCSSString = await InspectorFrontendHost.pickColorFromScreen().catch((error) => {
```
Comment (dcrousso) [inline Source/WebInspectorUI/UserInterface/Views/ColorPicker.js]: NIT: It's kinda weird IMO to mix `Promise` and `await` like this.
```
let pickedColorCSSString = null;
try {
    pickedColorCSSString = await InspectorFrontendHost.pickColorFromScreen();
} catch (e) {
    WI.reportInternalError(e);
}
```

File: Source/WebInspectorUI/UserInterface/Views/InlineSwatch.js [hunk 13]
```diff
@@ -192,10 +192,13 @@ WI.InlineSwatch = class InlineSwatch extends WI.Object
 
         switch (this._type) {
         case WI.InlineSwatch.Type.Color:
+            let title = WI.UIString("Click to select a color");
```
Comment (dcrousso) [inline Source/WebInspectorUI/UserInterface/Views/InlineSwatch.js]: NIT: I wonder if these should end in `"."` if there's multiple lines 🤔

File: Source/WebCore/inspector/InspectorFrontendHost.cpp [hunk 14]
```diff
@@ -494,6 +497,45 @@ void InspectorFrontendHost::load(const String& path, Ref<DeferredPromise>&& prom
...
+}
+
+void InspectorFrontendHost::pickColorFromScreen(Ref<DeferredPromise>&& promise)
+{
+    if (!m_client) {
+        promise->reject(InvalidStateError);
+        return;
+    }
+
+    if (!canPickColorFromScreen()) {
```
Comment (dcrousso) [inline Source/WebCore/inspector/InspectorFrontendHost.cpp]: NIT: normally we don't also check the `can*` inside the function that actually does the thing (the idea being that we expect callers to do that beforehand) as we should implement the function such that it "works" (i.e. doesn't crash) if we actually cannot do the thing, but i suppose it couldn't hurt

**Review by dcrousso — APPROVED** [review]
🤩🤩🤩

File: Source/WebCore/inspector/InspectorFrontendHost.cpp (hunk 1 above)
Comment (patrickangle) [reply Source/WebCore/inspector/InspectorFrontendHost.cpp:522]: We can serialize just fine, but in the frontend we currently can't deserialize into a `WI.Color` things like rec2020, etc. so until we implement those we need to convert them on the backend to another colorspace.

File: Source/WebInspectorUI/UserInterface/Views/ColorPicker.js (hunk 9 above)
Comment (patrickangle) [reply Source/WebInspectorUI/UserInterface/Views/ColorPicker.js:66]: The result of `WI.ImageUtilities.useSVGSymbol` is a `<div>` wrapping an `<svg>`. I've moved the sizing into the CSS for the element.

File: Source/WebInspectorUI/UserInterface/Views/ColorPicker.js [hunk 15]
```diff
@@ -68,6 +94,33 @@ WI.ColorPicker = class ColorPicker extends WI.Object
...
+
+    static async pickColorFromScreen({suggestedFormat, suggestedGamut, forceSuggestedFormatAndGamut} = {})
+    {
+        console.assert(InspectorFrontendHost.canPickColorFromScreen());
+
+        // There is a brief moment where the frontend page remains interactable before the backend actually begins the
+        // modal color picking mode. In order to avoid accidentally hovering an element and showing its highlight on the
+        // page and not being able to hide the highlight while selecting a color, make the document inert so that even
+        // immediate mouse movement doesn't accidentaly cause any highlighting to occur.
+        document.body.inert = true;
```
Comment (JosephPecoraro) [inline Source/WebInspectorUI/UserInterface/Views/ColorPicker.js:107]: Whoa. 🤯

File: Source/WebKit/UIProcess/Inspector/mac/WebInspectorUIProxyMac.mm [hunk 16]
```diff
@@ -590,6 +592,19 @@ - (void)inspectorViewController:(WKInspectorViewController *)inspectorViewContro
         completionHandler(nullString());
 }
 
+void WebInspectorUIProxy::platformPickColorFromScreen(CompletionHandler<void(const std::optional<WebCore::Color>&)>&& completionHandler)
+{
+    NSColorSampler *sampler = [[NSColorSampler alloc] init];
```
Comment (JosephPecoraro) [inline Source/WebKit/UIProcess/Inspector/mac/WebInspectorUIProxyMac.mm]: Is this file ARC? If not this should probably be put into a `RetainPtr` with `adoptNS` or manually released?

Comment (JosephPecoraro) [reply Source/WebKit/UIProcess/Inspector/mac/WebInspectorUIProxyMac.mm]: You can check `leaks <pid-of-web-process>` after using the color picker a few times to see if an `NSColorSampler` leaks.

Comment (patrickangle) [reply Source/WebKit/UIProcess/Inspector/mac/WebInspectorUIProxyMac.mm]: This was indeed leaking! `adoptNS` into a `RetainPtr` fixes this leak by allowing the initial allocation to be balanced with a release at the end of the function scope. The object is still retained until the callback is called, which is managed by the `NSColorSampler` API.

**Comment by yisibl** [comment]
Hi, @patrickangle Do you have plans to implement [EyeDropper API](https://github.com/WICG/eyedropper-api/)? This way the color picker can be easily called from JS in the web page.

Chrome 95 starts supporting this API and I use it in Chrome DevTools to implement the new color picker. https://bugs.chromium.org/p/chromium/issues/detail?id=1245191
